lib/phoenix/router/helpers.ex

```diff
@@ -9,20 +9,26 @@ defmodule Phoenix.Router.Helpers do
   @doc """
   Callback invoked by url generated in each helper module.
   """
-  def url(_router, %Conn{private: private}) do
-    private.phoenix_endpoint.url
+  def url(router, %Conn{private: %{phoenix_router_url: %URI{} = uri}}) do
+    url(router, uri)
+  end
+  def url(_router, %Conn{private: %{phoenix_router_url: "" <> url}}) do
+    url
+  end
+  def url(_router, %Conn{private: %{phoenix_endpoint: endpoint}}) do
+    endpoint.url()
   end
```

Comment:
Let's get all of those three clauses and put them in a case. Something like:

```elixir
def url(_router, %Conn{private: private}) do
  case private do
    ...
```

The reason is that the formatter won't keep those clauses semantically grouped as we have right now and the current approach may end-up doing multiple accesses on private at runtime.